rp2040-hal/src/pio.rs

```diff
@@ -0,0 +1,598 @@
+        self.sm().sm_execctrl.read().exec_stalled().bits()
+    }
+
+    fn block(&self) -> &rp2040_pac::pio0::RegisterBlock {
+        unsafe { &*self.block }
+    }
+
+    fn sm(&self) -> &rp2040_pac::pio0::SM {
+        &self.block().sm[self.id as usize]
+    }
+
+    // TODO: Other interrupt control functions
+    pub fn enable_interrupt_0(&self, enabled: bool) {
+        self.block().sm_irq[0]
+            .irq_inte
+            .modify(|_, w| match self.id {
+                0 => w.sm0().bit(enabled),
+                1 => w.sm1().bit(enabled),
+                2 => w.sm2().bit(enabled),
+                3 => w.sm3().bit(enabled),
+                _ => unreachable!(),
+            });
+    }
+
+    pub fn enable_tx_not_full_interrupt_1(&self, enabled: bool) {
+        self.block().sm_irq[1]
+            .irq_inte
+            .modify(|_, w| match self.id {
+                0 => w.sm0_txnfull().bit(enabled),
+                1 => w.sm1_txnfull().bit(enabled),
+                2 => w.sm2_txnfull().bit(enabled),
+                3 => w.sm3_txnfull().bit(enabled),
+                _ => unreachable!(),
+            });
+    }
+
+    pub fn enable_rx_not_empty_interrupt_1(&self, enabled: bool) {
+        self.block().sm_irq[1]
+            .irq_inte
+            .modify(|_, w| match self.id {
+                0 => w.sm0_rxnempty().bit(enabled),
+                1 => w.sm1_rxnempty().bit(enabled),
+                2 => w.sm2_rxnempty().bit(enabled),
+                3 => w.sm3_rxnempty().bit(enabled),
+                _ => unreachable!(),
+            });
+    }
+
+    pub fn read_rx(&self) -> Option<u32> {
+        let level = match self.id {
+            0 => self.block().flevel.read().rx0().bits(),
+            1 => self.block().flevel.read().rx1().bits(),
+            2 => self.block().flevel.read().rx2().bits(),
+            3 => self.block().flevel.read().rx3().bits(),
+            _ => unreachable!(),
+        };
+        if level > 0 {
+            Some(self.block().rxf[0].read().bits())
+        } else {
+            None
+        }
+    }
+}
+
+/// Comparison used for `mov x, status` instruction.
+#[derive(Debug, Clone, Copy)]
+pub enum MovStatusConfig {
+    /// The `mov x, status` instruction returns all ones if TX FIFO level is below the set status, otherwise all zeros.
+    Tx(u8),
+    /// The `mov x, status` instruction returns all ones if RX FIFO level is below the set status, otherwise all zeros.
+    Rx(u8),
+}
+
+/// Shift direction for input and output shifting.
+#[derive(Debug, Clone, Copy)]
+pub enum ShiftDirection {
+    /// Shift register to left.
+    Left,
+    /// Shift register to right.
+    Right,
+}
+
+impl ShiftDirection {
+    fn bit(self) -> bool {
+        match self {
+            Self::Left => false,
+            Self::Right => true,
+        }
+    }
+}
+
+/// Builder to deploy a fully configured PIO program on one of the state
+/// machines.
+#[derive(Debug)]
+pub struct PIOBuilder<'a> {
```

Comment:
It's a minor thing, but since this doesn't actually build some `PIO` struct, maybe we could name this just `PIO` or something different entirely, similar to the `Assembler` in pio-rs.